tests/laplace.hpp

```diff
@@ -0,0 +1,240 @@
+ * top-level directory                                                   *
+ *                                                                       *
+ * SPDX-License-Identifier: BSD-3-Clause                                 *
+ *************************************************************************/
+
+#include <deal.II/base/function.h>
+#include <deal.II/base/index_set.h>
+#include <deal.II/base/quadrature.h>
+#include <deal.II/distributed/tria.h>
+#include <deal.II/dofs/dof_handler.h>
+#include <deal.II/dofs/dof_tools.h>
+#include <deal.II/fe/fe_q.h>
+#include <deal.II/fe/fe_values.h>
+#include <deal.II/grid/filtered_iterator.h>
+#include <deal.II/grid/grid_generator.h>
+#include <deal.II/lac/constraint_matrix.h>
+#include <deal.II/lac/full_matrix.h>
+#include <deal.II/lac/trilinos_solver.h>
+#include <deal.II/lac/trilinos_sparse_matrix.h>
+#include <deal.II/lac/trilinos_sparsity_pattern.h>
+#include <deal.II/lac/vector.h>
+#include <deal.II/numerics/data_out.h>
+#include <deal.II/numerics/error_estimator.h>
+#include <deal.II/numerics/vector_tools.h>
+
+#include <boost/mpi.hpp>
+#include <string>
+
+template <int dim, typename VectorType>
+class Laplace
+{
+public:
+  Laplace(boost::mpi::communicator const &comm, unsigned int fe_degree);
+
+  void setup_system();
+
+  void assemble_system(dealii::Function<dim> const &source);
+
+  template <typename PreconditionerType>
+  VectorType solve(PreconditionerType &preconditioner);
+
+  double compute_error(dealii::Function<dim> const &exact_solution);
+
+  void output_results() const;
+
+  template <typename PreconditionerType>
+  void run(PreconditionerType &preconditioner,
+           dealii::Function<dim> const &source);
+
+private:
+  boost::mpi::communicator _comm;
+  dealii::parallel::distributed::Triangulation<dim> _triangulation;
+  dealii::FE_Q<dim> _fe;
+  dealii::DoFHandler<dim> _dof_handler;
+  dealii::IndexSet _locally_owned_dofs;
+  dealii::IndexSet _locally_relevant_dofs;
+  dealii::ConstraintMatrix _constraints;
+  dealii::TrilinosWrappers::SparseMatrix _system_matrix;
+  VectorType _locally_relevant_solution;
+  VectorType _system_rhs;
+};
+
+template <int dim, typename VectorType>
+Laplace<dim, VectorType>::Laplace(boost::mpi::communicator const &comm,
+                                  unsigned int fe_degree)
+    : _comm(comm), _triangulation(_comm), _fe(fe_degree),
+      _dof_handler(_triangulation)
+{
+}
+
+template <int dim, typename VectorType>
+void Laplace<dim, VectorType>::setup_system()
+{
+  dealii::GridGenerator::hyper_cube(_triangulation);
+  _triangulation.refine_global(3);
+
+  _dof_handler.distribute_dofs(_fe);
+
+  // Get the IndexSets
+  _locally_owned_dofs = _dof_handler.locally_owned_dofs();
+  dealii::DoFTools::extract_locally_relevant_dofs(_dof_handler,
+                                                  _locally_relevant_dofs);
+
+  // Resize the vectors
+  _locally_relevant_solution.reinit(_locally_owned_dofs, _locally_relevant_dofs,
```

Comment:
I'm not sure I understand why it takes 2 "maps" in.

Reply:
This is the only way to create a ghosted vector in deal.II. It allows you to know what is ghosted and what is not. I don't think this is necessary with Trilinos vectors but PETSc vectors behave very differently and we want to have only one interface to switch between all the type of vectors that are supported http://dealii.org/8.5.0/doxygen/deal.II/DEALGlossary.html#GlossGhostedVector